### DISCUSSION TITLE
Accessing multiple databases

### DISCUSSION THREAD

**Discussion by chimeracle** [reflex-dev/reflex#3217] [Q&A] [answered]
What would be the best way to go about accessing multiple databases? Currently, I'm using the db_url in the rxconfig file for the main database and passing the whole connection string for a second database in the rx.session(). Is there a cleaner, more secure way to do it with a second db_url or something in the rxconfig file?

**Comment by picklelo** [comment]
There's to a really a builtin way to do this as it would be hard to map the models to the corresponding database. If you need this, the way you're doing it seems like a good option. We're open to suggestions on how we can make the API nicer for these use cases.

**Reply by chimeracle** [reply]
Would it be possible to have multiple db_url variables in rxconfig and then reference the associated url in the table class, either inside rx.Model or as an arg to tie them together? The database url's could live in a .env and are called in rxconfig and for models. I suggest this knowing very little on how Reflex/Alembic handles linking these.



```
class TestModel(rx.Model, table=True, database=db_url):
    id: Optional[int] = Field(primary_key=True)
    field1: int
    field2: bool

class TestModel2(rx.Model(db_url2), table=True):
    id: Optional[int] = Field(primary_key=True)
    field1: int
    field2: str
```

**Reply by picklelo** [reply, marked as answer]
@chimeracle Currently the models aren't really tied to a database, it's the sessions themselves. Could you use your original approach with a helper function to achieve this? I think it should be pretty clean:

```python
def session2():
    return rx.session(db_url2)

...

with session2() as session:
   ...
```

**Reply by chimeracle** [reply]
@picklelo Thanks for the suggestion. I'll use it in the base/parent state.

This may be better as a new discussion, but it pertains to this as I'm juggling a couple database connections. How can I safely store and use database and other credentials without exposing them to the client/user? Is it as simple as declaring them outside the state or can they be inside the state with a "_" in front denoting a backend variable?

**Reply by picklelo** [reply]
Only your UI code and state vars are sent to the client. As long as you don't put your DB URL in the UI itself, and don't make it a state var, it will stay private on the server. Using a backend var also works, it won't be sent to the client.

**Comment by benedikt-bartscher** [comment]
Related project: https://github.com/febus982/sqlalchemy-bind-manager
